### ISSUE TITLE
[ERTP] Move away from generator-based composition of mint configuration

### ISSUE THREAD

**Issue by DavidBruant** [Agoric/agoric-sdk#375]
Follow-up to https://github.com/Agoric/agoric-sdk/issues/80#issuecomment-561240691

Probably based on layer-cake after https://github.com/agoric-labs/layer-cake/pull/4 is merged

**Comment by DavidBruant** [comment]
I'm hitting a problem here when trying to refactor `makeMint` configuration so that it's based on trait layer-cake and i would like to discuss it here before making a decision

In the [current version of the code](https://github.com/Agoric/agoric-sdk/blob/b133ab0557af00edc960f66a0df87a366ecf2393/packages/ERTP/core/mint.js), the `make*Trait` function are called with the base/core layer as first argument. And also with additional arguments that are different for the different methods

```js
makeAssayTrait(coreAssay)
makePaymentTrait(corePayment, assay);
makePurseTrait(corePurse, assay);
makeMintTrait(coreMint, assay, unitOps, mintKeeper);
```

and each `make*Trait` function is defined with this pattern (before move to layer-cake):
```js
function* make*Trait(core*, ...otherArgs){
  yield {
    // trait
  }
}
```

I see 2 different problems here:
1) layer-cake expects a function that only takes (i'm simplifying) the base layer as argument and no additional arguments
2) the additional arguments for each method feel a bit ad-hoc

I think problem 1) can be solved by moving to a pattern like (after move to generator-less version):
```js
function make*Trait(...otherArgs){
  return core* => ({
    // trait
  })
}
```

However, I'm not sure what to do with 2). I haven't been doing the proper code archeology but my gut feeling is that for various customization use cases (probably mainly pixelConfig), there has been a need to provide various arguments for the purpose of making useful customizations. And that's the reason different `make*Trait` methods need different things right now.

But i feel like people making their own assay may want to provide methods based on the `unitsOps` or maybe the `mint` being built like `makeEmptyPurse` (from `coreAssay`) currently does
Overall, i think it's currently very hard/impossible for custom methods to reproduce what core methods can do because of a lack of access to internal variables in `makeMint` 

My proposal is:
replace all `...otherArgs` by a single object argument pass to all customization methods:
```js
function make*Trait(makeMintContext){
  return core* => ({
    // trait
  })
}
```

The properties of `makeMintContext` are lots of internal things defined in `makeMint`. Right now, I imagine the following object (based on what's used internally by core methods):
`{assay, mint, unitOps, mintKeeper, paymentKeeper, purseKeeper, takePayment, takePaymentAndKill, depositInto }`

because some of these don't exist for some methods (like `assay` is irrelevant in `makeAssayTrait`), i imagine at least some of the properties will be getters and return `undefined` when appropriate

**Comment by katelynsills** [comment]
Yup, the arguments were driven by the pixel configuration. I like the idea of providing a context object. The only problem I see is that given how only certain things will be made at certain times, it's probably not possible to provide the same context object for all of the functions. 

There's a simplification of ERTP in the early design process that will move nearly all of the methods from payments and purses to the assay, and drops any reference to the assay in the payments and purses. This may make what you're doing now a lot easier, since there won't be a need to add use objects to purses and payments, and there will be a lot fewer interconnected references between mints/assays/purses/payments. It's still in the early stages, so I can't say for sure how the gallery pixel demo would change, but I think it would probably change significantly.

**Comment by katelynsills** [comment]
Closed by #389
